compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -431,7 +431,21 @@ fn layout_of_uncached<'tcx>(
                 .size
                 .checked_mul(e_len, dl)
                 .ok_or_else(|| error(cx, LayoutError::SizeOverflow(ty)))?;
-            let align = dl.vector_align(size);
+
+            let (abi, align) = if def.repr().packed() && !e_len.is_power_of_two() {
+                // Non-power-of-two vectors have padding up to the next power-of-two.
```

Comment:
iirc LLVM doesn't guarantee vectors don't have more padding than just to the next power of two, e.g. padding `<2 x i8>` to 128-bits.

Reply:
https://llvm.org/docs/LangRef.html#vector-type seems to say that LLVM guarantees that vectors have no padding in their in memory representation? Padding `<2 x i8>` to 128 bits is something that LLVM does inside the vector registers (e.g. on aarch64), but never in memory.

Reply:
[`core::mem::size_of`](https://doc.rust-lang.org/core/mem/fn.size_of.html):
> More specifically, this is the offset in bytes between successive elements in an array with that item type including alignment padding.

Reply:
Whoops, I meant rustc assumes that npot vectors round up to the next power of two.

Reply:
> > that's only for load/store, LLVM vector types still have big enough alignment that Rust requires padding in many cases.
> 
> I thought LLVM types didn't have any sort of intrinsic alignment? And even if they have a preferred alignment that's too big, can't you just always annotate load/store instructions with the smaller alignment?

they do have an intrinsic ABI alignment, it's used for default alignment of load/store/alloca when not explicitly specified and for stack spilling and a few other things

Reply:
> I'm not sure if it's guaranteed, but rustc does make the assumption (somewhere, I forget where)

just found this:
https://lang-team.rust-lang.org/frequently-requested-changes.html#size--stride
> Rust assumes that the size of an object is equivalent to the stride of an object